synapse/federation/federation_client.py

```diff
@@ -1364,13 +1364,59 @@ async def send_request(
 
             return room, children, inaccessible_children
 
-        # TODO Fallback to the old federation API and translate the results.
-        return await self._try_destination_list(
-            "fetch room hierarchy",
-            destinations,
-            send_request,
-            failover_on_unknown_endpoint=True,
-        )
+        try:
+            return await self._try_destination_list(
+                "fetch room hierarchy",
+                destinations,
+                send_request,
+                failover_on_unknown_endpoint=True,
+            )
+        except SynapseError as e:
+            # Fallback to the old federation API and translate the results if
+            # no servers implement the new API.
+            #
+            # The algorithm below is a bit inefficient as it only attempts to
+            # get information for the requested room, but the legacy API may
+            # return additional layers.
+            if e.code == 502:
```

Comment:
Does Synapse not return a `400, M_UNRECOGNIZED` for endpoints it fails to understand, rather than a 502?

Reply:
This is from https://github.com/matrix-org/synapse/blob/4b1722a32e9d6c60cee83c6d9ed769689e493ce8/synapse/federation/federation_client.py#L600

Reply:
aha, thanks!

Reply:
The goal is only to run this if we were unable to fetch data from *all* destinations, not just if we were unable to fetch it from the first destination. (This is different than our fallback of v2 -> v1 for e.g. `send_join`, which we do on a per-destination basis).